charts/generic-service/templates/controller.yaml

```
@@ -51,59 +58,56 @@ spec:
-        {{- else }}
-          {{ fail "Ingress must use Istio or nginx if rollout.strategy is Canary" }}
+      {{- with .Values.rollout.analysis -}}
+        analysis: {{- . | toYaml | nindent 8 }}
+      {{- end }}
+      {{- if .Values.rollout.useTrafficRouting -}}
+        {{- if not eq .Values.rollout.strategy "Canary" }}
+          {{ fail "Traffic routing is only available if rollout.strategy is Canary" }}
+        {{- end }}
+        stableService: {{ include "generic-service.fullname" . }}
+        stableMetadata:
+          labels:
+            role: stable
+        canaryService: {{ include "generic-service.fullname" . }}-preview
+        canaryMetadata:
+          labels:
+            role: preview
+        trafficRouting:
+          {{- if .Values.ingress.istio.enabled }}
+            istio:
+              virtualService:
+                name: {{ include "generic-service.fullname" . }}
+          {{- else if eq .Values.ingress.class "nginx" }}
+            nginx:
+              stableIngress: {{ include "generic-service.fullname" . }}
+          {{- else }}
+            {{ fail "Ingress must use Istio or nginx if rollout.strategy is Canary" }}
+          {{- end }}
+        {{- end }}
+        {{- with .Values.rollout.steps }}
+        steps: {{- . | toYaml | nindent 8 }}
         {{- end }}
-    {{- else if not .Values.rollout.flagger }}
-      {{ fail "rollout.flagger must be true or rollout.controller must be ArgoRollout if rollout.strategy is Canary" }}
     {{- end }}
   {{- else if eq .Values.rollout.strategy "BlueGreen" }}
-    {{- if not .Values.ingress.enabled }}
-      {{ fail "ingress.enabled must be true if rollout.strategy is BlueGreen" }}
-    {{- end }}
     {{- if eq .Values.rollout.controller "ArgoRollout" }}
-    blueGreen:
-      activeService: {{ include "generic-service.fullname" . }}
-      activeMetadata:
-        labels:
-          role: stable
-      previewService: {{ include "generic-service.fullname" . }}-preview
-      previewMetadata:
-        labels:
-          role: preview
-      autoPromotionEnabled: {{ .Values.rollout.autoPromotion }}
-      {{- if .Values.rollout.analysis }}
-      prePromotionAnalysis: {{- .Values.rollout.analysis | toYaml | nindent 8 }}
-      {{- end }}
-    {{- else if not .Values.rollout.flagger }}
-      {{ fail "rollout.flagger must be true or rollout.controller must be ArgoRollout if rollout.strategy is BlueGreen" }}
+      blueGreen:
+        activeService: {{ include "generic-service.fullname" . }}
+        activeMetadata:
+          labels:
+            role: stable
+        previewService: {{ include "generic-service.fullname" . }}-preview
+        previewMetadata:
+          labels:
+            role: preview
+        autoPromotionEnabled: {{ .Values.rollout.autoPromotion }}
+        {{- if .Values.rollout.analysis }}
+          prePromotionAnalysis: {{- .Values.rollout.analysis | toYaml | nindent 8 }}
+        {{- end }}
```

Indent seems to be too deep. Would put `prePromotionAnalysis` under `autoPromotionEnabled`.

```suggestion
        {{- if .Values.rollout.analysis }}
        prePromotionAnalysis: {{- .Values.rollout.analysis | toYaml | nindent 10 }}
        {{- end }}
```

---

If I understood you correctly, made `prePromotionAnalysis` dependent on whether `autoPromotionEnabled` is `true`.